lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -362,10 +409,41 @@ private TaskStateEvent GetTaskErrorEventByExceptionType(IFailedTask failedTask)
             {
                 return TaskStateEvent.FailedTaskCommunicationError;
             }
-            else
+            if (exception is IMRUTaskSystemException)
             {
                 return TaskStateEvent.FailedTaskSystemError;
             }
+
+            // special case for communication error during group communication initialization
+            if (exception is TaskClientCodeException)
+            {
+                {
+                    // try extract cause and check whether it is InjectionException for GroupCommClient
+                    if (exception.InnerException != null &&
+                        exception.InnerException is InjectionException &&
+                        exception.InnerException.Message.Contains("GroupCommClient"))
+                    {
+                        Logger.Log(Level.Info, "GetTaskErrorEventByExceptionType:FailedTaskCommunicationError with task id {0}", failedTask.Id);
+                        return TaskStateEvent.FailedTaskCommunicationError;
+                    }
+                }
+            }
+
+            if (exception is AggregateException)
+            {
+                {
+                    // try extract cause and check whether it is InjectionException for GroupCommClient
+                    if (exception.InnerException != null &&
+                        exception.InnerException is IOException)
+                    {
+                        Logger.Log(Level.Info, "GetTaskErrorEventByExceptionType:AggregateException-IOException with task id {0}", failedTask.Id);
+                        return TaskStateEvent.FailedTaskCommunicationError;
+                    }
+                }
+            }
```

Comment:
This is how we handle the IMRU task exceptions:

All the known exceptions caught in Task Host Call() method should be converted to predefined exceptions like IMRUTaskAppException, IMRUTaskGroupCommunicationException or IMRUTaskSystemException so that in TaskManager, we only check those exceptions. 

For exceptions are thrown in task host constructor, it is controlled by TaskRuntime which is wrapped in TaskClientCodeException with InjectionException as inner exception. This is in REEF core. We need a better way to distinguish details of those exceptions. Can be done in separate Jira. 

For unknown exceptions thrown with FailedTask, we default to IMRUTaskSystemException so that it is still recoverable.